File: tools/copyrighter/copyright.rb
require_relative 'copyright_param'

class Copyright

  attr_reader :text, :params

  def initialize(text)

    @text = text
    @params = []

  end

  def add_param(param)
    @params.push param
  end

  def set_param(name, value)

    @params.each do |param|
      if name == param.name
        param.value = value
        return
      end
    end

    add_param CopyrightParam.new(name, value)

  end

  def process

    new_text = text

    @params.each do |param|
      name = param.decorated_name
      value = param.value

      new_text = new_text.sub name, value

    end

    return new_text

  end

  def self.remove_header_comment(text)

    lines = text.lines "\n"

    index = find_first_non_comment_line_index lines
    return text if index == 0

    lines = lines[index..lines.length]
    return lines.join

  end

  def self.find_first_non_comment_line_index(lines)
    comment_found = false
    multiline = false
    skip_blank_lines = false

    lines.each_with_index do |line, index|

      stripped_line = line.strip

      if skip_blank_lines
        return index if stripped_line.length > 0
        next
      end

      if comment_found
        skip_blank_lines = multiline ? stripped_line.end_with?('*/') : !stripped_line.start_with?('//')
      else # comment not found
        multiline = stripped_line.start_with? '/*'
        comment_found = multiline || stripped_line.start_with?('//')

        unless comment_found
          return index if stripped_line.length > 0
          skip_blank_lines = true
        end
      end

    end

    return 0 # no comment found
  end

end

File: tools/copyrighter/copyright.py
from copyright_param import CopyrightParam


class Copyright:

    def __init__(self, text):
        self.text = text
        self.params = []

    def add_param(self, param):
        self.params.append(param)

    def set_param(self, name, value):
        for param in self.params:
            if name == param.name:
                param.value = value
                return

        self.add_param(CopyrightParam(name, value))

    def process(self):
        new_text = self.text

        for param in self.params:
            name = param.decorated_name
            value = param.value

            new_text = new_text.replace(name, value)

        return new_text

    @staticmethod
    def remove_header_comment(text):
        lines = text.splitlines(keepends=True)

        index = Copyright.find_first_non_comment_line_index(lines)
        if index == 0:
            return text

        lines = lines[index:]
        return ''.join(lines)

    @staticmethod
    def find_first_non_comment_line_index(lines):
        comment_found = False
        multiline = False
        skip_blank_lines = False

        for index, line in enumerate(lines):
            stripped_line = line.strip()

            if skip_blank_lines:
                if len(stripped_line) > 0:
                    return index
                continue

            if comment_found:
                if multiline:
                    skip_blank_lines = stripped_line.endswith('*/')
                else:
                    skip_blank_lines = not stripped_line.startswith('//')
            else:
                multiline = stripped_line.startswith('/*')
                comment_found = multiline or stripped_line.startswith('//')

                if not comment_found:
                    if len(stripped_line) > 0:
                        return index
                    skip_blank_lines = True

        return 0  # no comment found
